unified_planning/io/anml_reader.py

```diff
@@ -0,0 +1,810 @@
+        if effect_res[0] == TK_WHEN:  # Conditional effect
+            condition_interval = effect_res[1][0]
+            condition_exp_res = effect_res[1][1]
+            effect_interval = effect_res[2][0]
+            effect_exp = effect_res[2][1][0]
+            if str(condition_interval) != str(effect_interval):
+                raise UPProblemDefinitionError(
+                    "In conditional effect parsing the "
+                    + "condition time interval and the effect time interval are different, "
+                    + "this is not supported by the UP."
+                )
+            if len(interval_res) > 0:
+                raise UPProblemDefinitionError(
+                    "A conditional effect can not have an interval specified before the when keyword."
+                )
+            condition = self._parse_expression(condition_exp_res, parameters, types_map)
+            interval_res = condition_interval
+        else:
+            condition = self._em.TRUE()
+            effect_exp = effect_res[0]
+        up_interval = self._parse_interval(interval_res, types_map, is_global)
+        fluent_ref = effect_exp[0]
+        assignment_operator = effect_exp[1]
+        assigned_expression = effect_exp[2]
+        up_fluent = self._parse_expression(fluent_ref, parameters, types_map)
+        assert (
+            up_fluent.is_fluent_exp()
+        ), "left side of the assignment is not a valid fluent"
+        up_value = self._parse_expression(assigned_expression, parameters, types_map)
+        if assignment_operator == TK_ASSIGN:
+            kind = EffectKind.ASSIGN
+        elif assignment_operator == TK_INCREASE:
+            kind = EffectKind.INCREASE
+        elif assignment_operator == TK_DECREASE:
+            kind = EffectKind.DECREASE
+        else:
+            raise NotImplementedError(
+                f"Currently the unified planning does not support the assignment operator {assignment_operator}"
+            )
+        assert isinstance(
+            up_interval, Timing
+        ), "An effect with a durative interval is not supported"
+        return (up_interval, Effect(up_fluent, up_value, condition, kind=kind))
+
+    def _parse_expression(
+        self,
+        expression: Union[ParseResults, List, str],
+        parameters: Dict[str, "up.model.Parameter"],
+        types_map: Dict[str, "Type"],
+    ) -> "up.model.FNode":
+        # A string here means it is a number or a boolean token. To avoid code duplication, just
+        # wrap it in a temporary list and let the stack management handle it.
+        if isinstance(expression, str):
+            expression = [expression]
+        stack: List[Tuple[Union[ParseResults, List], bool, Dict[str, "Variable"]]] = [
+            (expression, False, {})
+        ]
+        solved: List[FNode] = []
+        while 1:
+            try:
+                exp, already_expanded, vars = stack.pop()
+            except IndexError:
+                break
```

Comment:
Why not checking the stack size in the while condition?

Reply:
I thought that the `len` time complexity was O(n), but I checked and it actually is O(1).